generate_metap_agg.py

```diff
@@ -1,191 +1,454 @@
+        """
+        # Get list of workflow IDs that have already been processed
+        mp_wf_in_agg = self.get_previously_aggregated_workflow_ids()
+
+        # Get list of all workflow records
+        mp_wf_recs = self.get_workflow_records()
+
+        # If there are any records that were not processed, return FALSE
+        check = [x["id"] in mp_wf_in_agg for x in mp_wf_recs]
+        if all(check):
+            return True
+        else:
+            return False
+
+    @abstractmethod
+    def process_activity(self, act):
+        """
+        Abstract method to process an activity record.  This method must be implemented in the subclass.
+
+        Parameters
+        ----------
+        act : dict
+            Activity record to process
+
+        Returns
+        -------
+        dict
+            Dictionary of functional annotations with their respective counts
+        """
+        pass
+
+
+class MetaProtAgg(Aggregator):
     """
-    MetaP Aggregation class
+    Metaproteomics Aggregator class
+
+    Parameters
+    ----------
+    dev : bool
+        Flag to indicate if production or development API should be used
+        Default is True, which uses the development API
+
+    Attributes
+    ----------
+    base_url : str
+        Base URL for the API
+    nmdc_api_token : str
+        API token to access the API
+
+    Notes
+    -----
+    This class is used to aggregate functional annotations from metaproteomics workflows in the NMDC database.
     """
-    _BASE_URL_ENV = "NMDC_BASE_URL"
-    _base_url = "https://data.microbiomedata.org/data"
-    _BASE_PATH_ENV = "NMDC_BASE_PATH"
-    _base_dir = "/global/cfs/cdirs/m3408/results"
 
     def __init__(self):
-        url = os.environ["MONGO_URL"]
-        client = MongoClient(url, directConnection=True)
-        self.db = client.nmdc
-        self.agg_col = self.db.metap_gene_function_aggregation
-        self.act_col = self.db.workflow_execution_set 
-        self.do_col = self.db.data_object_set
-        self.base_url = os.environ.get(self._BASE_URL_ENV, self._base_url)
-        self.base_dir = os.environ.get(self._BASE_PATH_ENV, self._base_dir)
-
-    def get_kegg_terms(self, url, gene_list):
-        kos = {}
-        fn = url.replace(self.base_url, self.base_dir)
-
-        if os.path.exists(fn):
-            lines = open(fn)
-        else:
-            s = requests.Session()
-            resp = s.get(url, headers=None, stream=True)
-            if not resp.ok:
-                raise OSError(f"Failed to read {url}")
-            lines = resp.iter_lines()
-
-        for line in lines:
-            if isinstance(line, bytes):
-                line = line.decode()
-            anno = AnnotationLine(line, gene_list)
-            if anno.kegg:
-                kos[anno.id] = anno.kegg
-        return kos
-
-    def find_anno(self, dos):
+        super().__init__()
+        self.aggregation_filter = '{"was_generated_by":{"$regex":"^nmdc:wfmp"}}'
```

Comment:
I changed the default regex here to be a so-called "prefix expression" (<https://www.mongodb.com/docs/v6.0/reference/operator/query/regex/#index-use>):

> Further optimization can occur if the regular expression is a "prefix expression", which means that all potential matches start with the same string. This allows MongoDB to construct a "range" from that prefix and only match against those values from the index that fall within that range.

> A regular expression is a "prefix expression" if it starts with a caret (^) or a left anchor (\A), followed by a string of simple symbols. For example, the regex /^abc.*/ will be optimized by matching only against the values from the index that start with abc.

Please update the relevant docstrings to reflect this. We should encourage this optimization for large collections.